syntax = "proto3";
package iroha.protocol;

import "primitive.proto";

message GetAccount {
  string account_id = 1;
}

message GetSignatories {
  string account_id = 1;
}

message GetAccountTransactions {
  string account_id = 1;
}

message GetAccountAssetTransactions {
  string account_id = 1;
  string asset_id = 2;
}

message Pager {
  bytes tx_hash = 1;
  uint32 limit = 2;
}

message GetAccountTransactionsWithPager {
  string account_id = 1;
  Pager pager = 2;
}

message GetAccountAssetsTransactionsWithPager {
  string account_id = 1;
  repeated string assets_id = 2;
  Pager pager = 3;
}

message GetAccountAssets {
  string account_id = 1;
  string asset_id = 2;
}

message GetAssetInfo {
  string asset_id = 1;
}

message GetRoles {

}

message GetRolePermissions{
  string role_id = 1;
}


message Query {
  message Payload {
    uint64 created_time = 1;
    string creator_account_id = 2;
     oneof query {
       GetAccount get_account = 3;
       GetSignatories get_account_signatories = 4;
       GetAccountTransactions get_account_transactions = 5;
       GetAccountAssetTransactions get_account_asset_transactions = 6;
       GetAccountTransactionsWithPager get_account_transactions_with_pager = 7;
       GetAccountAssetsTransactionsWithPager get_account_assets_transactions_with_pager = 8;
       GetAccountAssets get_account_assets = 9;
       GetRoles get_roles = 10;
       GetAssetInfo get_asset_info = 11;
       GetRolePermissions get_role_permissions = 12;
     }
     // used to prevent replay attacks.
     uint64 query_counter = 13;
  }

  Payload payload = 1;
  Signature signature = 2;
}
